core/primitives/src/errors.rs

```diff
@@ -26,74 +79,165 @@ impl std::fmt::Display for StorageError {
+    InvalidSignerId { signer_id: AccountId },
+    /// TX signer_id is not found in a storage
+    SignerDoesNotExist { signer_id: AccountId },
+    /// Transaction nonce must be account[access_key].nonce + 1
+    InvalidNonce { tx_nonce: Nonce, ak_nonce: Nonce },
+    /// TX receiver_id is not in a valid format or not satisfy requirements see `near_core::primitives::utils::is_valid_account_id`
+    InvalidReceiverId { receiver_id: AccountId },
+    /// TX signature is not valid
     InvalidSignature,
-    NotEnoughBalance(AccountId, Balance, Balance),
-    RentUnpaid(AccountId, Balance),
+     /// Account does not have enough balance to cover TX cost
+    NotEnoughBalance {
+        signer_id: AccountId,
+        #[serde(with = "u128_dec_format")]
+        balance: Balance,
+        #[serde(with = "u128_dec_format")]
+        cost: Balance,
+    },
+    /// Signer account rent is unpaid
+    RentUnpaid {
+        /// An account which is required to pay the rent
+        signer_id: AccountId,
+        /// Required balance to cover the state rent
+        #[serde(with = "u128_dec_format")]
+        amount: Balance,
+    },
+    /// An integer overflow occurred during transaction cost estimation.
     CostOverflow,
+    /// Transaction parent block hash doesn't belong to the current chain
     InvalidChain,
+    /// Transaction has expired
     Expired,
 }
 
-#[derive(BorshSerialize, BorshDeserialize, Serialize, Debug, Clone, PartialEq, Eq)]
+#[derive(
+    BorshSerialize, BorshDeserialize, Debug, Clone, PartialEq, Eq, Deserialize, Serialize, RpcError,
+)]
+#[rpc_error_variant = "InvalidAccessKey"]
 pub enum InvalidAccessKeyError {
-    AccessKeyNotFound(AccountId, PublicKey),
-    ReceiverMismatch(AccountId, AccountId),
-    MethodNameMismatch(String),
+    /// The access key identified by the public_key doesn't exist for the account
+    AccessKeyNotFound { account_id: AccountId, public_key: PublicKey },
+    /// Transaction receiver_id doesn't match the access key receiver_id
+    ReceiverMismatch { tx_receiver: AccountId, ak_receiver: AccountId },
+    /// Transaction method name isn't allowed by the access key
+    MethodNameMismatch { method_name: String },
+    /// The used access key requires exactly one FunctionCall action
     ActionError,
-    NotEnoughAllowance(AccountId, PublicKey, Balance, Balance),
+    /// Access Key does not have enough allowance to cover transaction cost
+    NotEnoughAllowance {
+        account_id: AccountId,
+        public_key: PublicKey,
+        #[serde(with = "u128_dec_format")]
+        allowance: Balance,
+        #[serde(with = "u128_dec_format")]
+        cost: Balance,
+    },
+}
+
+/// An error happened during Acton execution
+#[derive(
+    BorshSerialize, BorshDeserialize, Debug, Clone, PartialEq, Eq, Deserialize, Serialize, RpcError,
+)]
+#[rpc_error_variant = "ActionError"]
+pub struct ActionError {
+    pub index: Option<u64>,
```

Comment:
what is the index here? Why it's option?

Reply:
a doc comment added